src/hevm/src/EVM/ABI.hs

```diff
@@ -537,6 +540,28 @@ listP parser = between (char '[') (char ']') ((do skipSpaces
                                                   return a) `sepBy` (char ','))
 
 
+-- TODO: teach this to handle complex types for symbolic buffers
+decodeBuffer :: [AbiType] -> Buffer -> Maybe (Either [AbiValue] [SWord 256])
+decodeBuffer argTypes (ConcreteBuffer b)
+  = case runGetOrFail
+      (getAbiSeq (length argTypes) argTypes)
+      (BSLazy.fromStrict b) of
+    Right ("", _, args) -> Just $ Left $ toList args
+    _ -> Nothing
+decodeBuffer argTypes (SymbolicBuffer b)
+  | containsComplexTypes argTypes = Nothing
+  | otherwise = Just . Right $
+      fmap (\i -> fromBytes $ take 32 (drop (i*32) b)) [0..((length b) `div` 32 - 1)]
+
+containsComplexTypes :: [AbiType] -> Bool
+containsComplexTypes [] = False
+containsComplexTypes (hd : tl) = case hd of
+  AbiBytesDynamicType -> True
+  AbiStringType -> True
+  AbiArrayDynamicType _ -> True
+  AbiArrayType _ _ -> True
+  AbiTupleType _ -> True
```

Comment:
tuples are only complex if their arguments are; `AbiTupleType [a,b]` is equivalent to `[a, b]`

Reply:
Also, this function is actually just a copy of `abiKind a == Dynamic`